pynicotine/userlist.py

```diff
@@ -51,26 +74,106 @@ def add_user(self, user):
         self.core.watch_user(user, force_update=True)
 
         # Request user country
-        self.set_user_country(user, self.core.get_user_country(user))
+        if new_user:
+            self.set_user_country(user, self.core.get_user_country(user))
 
     def remove_user(self, user):
+
+        if user in self.buddies:
+            del self.buddies[user]
+
+        self.save_user_list()
+
         if self.ui_callback:
             self.ui_callback.remove_user(user)
 
-    def save_user_list(self, user_list):
-        config.sections["server"]["userlist"] = user_list
-        config.write_configuration()
+    def set_user_note(self, user, note):
 
-    def get_user_status(self, msg):
-        """ Server code: 7 """
+        if user not in self.buddies:
+            return
+
+        self.buddies[user][1] = note
+        self.save_user_list()
 
         if self.ui_callback:
-            self.ui_callback.get_user_status(msg)
+            self.ui_callback.set_user_note(user, note)
+
+    def set_user_trusted(self, user, trusted):
+
+        if user not in self.buddies:
+            return
+
+        self.buddies[user][2] = trusted
+        self.save_user_list()
+
+        if self.ui_callback:
+            self.ui_callback.set_user_trusted(user, trusted)
+
+    def set_user_notify(self, user, notify):
+
+        if user not in self.buddies:
+            return
+
+        self.buddies[user][3] = notify
+        self.save_user_list()
+
+        if self.ui_callback:
+            self.ui_callback.set_user_notify(user, notify)
+
+    def set_user_prioritized(self, user, prioritized):
+
+        if user not in self.buddies:
+            return
+
+        self.buddies[user][4] = prioritized
+        self.save_user_list()
+
+        if self.ui_callback:
+            self.ui_callback.set_user_prioritized(user, prioritized)
+
+    def set_user_last_seen(self, user, online):
+
+        if user not in self.buddies:
+            return
+
+        previous_last_seen = self.buddies[user][5]
+
+        if online:
+            self.buddies[user][5] = ""
+            self.save_user_list()
+
+        elif not previous_last_seen:
+            self.buddies[user][5] = time.strftime("%m/%d/%Y %H:%M:%S")
+            self.save_user_list()
+
+        else:
+            return
+
+        if self.ui_callback:
+            self.ui_callback.set_user_last_seen(user, online)
 
     def set_user_country(self, user, country_code):
+
+        if user not in self.buddies:
+            return
+
+        self.buddies[user][6] = country_code
+        self.save_user_list()
+
         if self.ui_callback:
             self.ui_callback.set_user_country(user, country_code)
 
+    def save_user_list(self):
+        config.sections["server"]["userlist"] = list(self.buddies.values())
+
```

Comment:
`        config.write_configuration()`

Otherwise, newly added buddies are not immediately saved.